### PR TITLE
patch timing attack vulnerability ( #2174 )

### PR DESCRIPTION
**Pull request by divinity76**
replaces memcmp() in WardenMac::HandleHashResult with a constant-time check, see #2174 for details

##### CHANGES PROPOSED:
replaces memcmp() in WardenMac::HandleHashResult with a constant-time check to fix a timing attack vulnerability

##### ISSUES ADDRESSED:
https://github.com/azerothcore/azerothcore-wotlk/issues/2174

Closes 
https://github.com/azerothcore/azerothcore-wotlk/issues/2174

##### TESTS PERFORMED:
no tests performed, i have not even compiled it. hope someone else can do that

##### HOW TO TEST THE CHANGES:
?

***** IMPORTANT: *****
?

##### KNOWN ISSUES AND TODO LIST:

- [ ]
- [ ] 

##### Target branch(es):

Master

 
 

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by pklloveyou** [comment]
How to test?
TIM截图20190804201724

**Comment by divinity76** [comment]
> How to test?

#1: does it compile?

#2: .. i don't really know how to trigger this code, anyone?

**Comment by Motorheadx** [comment]
we have it is not possible to call as the warden does not work under OSX in AzerothCore

**Comment by divinity76** [comment]
> we have it is not possible to call as the warden does not work under OSX in AzerothCore
> https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Server/WorldSession.cpp#L1331

speaking of, the Win-version is also vulnerable, but this PR does not address the windows version, see https://github.com/azerothcore/azerothcore-wotlk/blob/c15206fc152a547dc39e50e8964738f8f115747a/src/server/game/Warden/WardenWin.cpp#L164

**Comment by pklloveyou** [comment]
{
        int result=0;
        const uint8_t *bufstart=(uint8_t*)(buff.contents() + 1);
        const uint8_t *digest= Module.ClientKeySeedHash;
        for (int i = 0; i < 20; ++i)
        {
            result |= bufstart[i] ^ digest[i];
        }
        if (result != 0)
        {
            Penalty();
            return;
        }
    }

**Comment by pklloveyou** [comment]
I was thinking, maybe I am not a hacker, so I can't simulate the attack to test this RP.

File: src/server/game/Warden/WardenMac.cpp [hunk 1]
```diff
@@ -161,13 +161,25 @@ void WardenMac::HandleHashResult(ByteBuffer &buff)
     //const uint8 validHash[20] = { 0x56, 0x8C, 0x05, 0x4C, 0x78, 0x1A, 0x97, 0x2A, 0x60, 0x37, 0xA2, 0x29, 0x0C, 0x22, 0xB5, 0x25, 0x71, 0xA0, 0x6F, 0x4E };
 
     // Verify key
-    if (memcmp(buff.contents() + 1, sha1.GetDigest(), 20) != 0)
+    // do not optimize key verification for speed, unless you really know what you are doing. 
+    // (you may introduce a timing attack vulnerability, 
+    // see https://github.com/azerothcore/azerothcore-wotlk/issues/2174 )
     {
```
Comment (Viste) [inline src/server/game/Warden/WardenMac.cpp]: you can delete {  or comment it

File: src/server/game/Warden/WardenMac.cpp [hunk 2]
```diff
@@ -161,13 +161,25 @@ void WardenMac::HandleHashResult(ByteBuffer &buff)
...
 #if defined(ENABLE_EXTRAS) && defined(ENABLE_EXTRA_LOGS)
-        sLog->outDebug(LOG_FILTER_WARDEN, "Request hash reply: failed");
+            sLog->outDebug(LOG_FILTER_WARDEN, "Request hash reply: failed");
 #endif
-        Penalty();
-        return;
+            Penalty();
+            return;
+        }
     }
```
Comment (Viste) [inline src/server/game/Warden/WardenMac.cpp]: and here

File: src/server/game/Warden/WardenMac.cpp (hunk 1 above)
Comment (divinity76) [reply src/server/game/Warden/WardenMac.cpp]: @Viste can, yes, but the { } creates a new variable scope, and inside is 3 new variables that are no longer relevant outside of the { } (`result`, `bufstart`, and `digest` are destroyed at the end of the `}` )

**Comment by BarbzYHOOL** [comment]
yes, mac warden never been implemented

**Comment by Kitzunu** [comment]
any update?

**Pull request closed without merging**